spec/p2p/reactor/p2p-api.md

```diff
@@ -0,0 +1,276 @@
+
+The above list is not exhaustive as it does not include all the `Switch` methods
+invoked by the PEX reactor, a special component that should be considered part
+of the p2p layer. This document does not cover the operation of the PEX reactor
+as a connection manager.
+
+### Peers State
+
+The first two methods in the switch API allow reactors to query the state of
+the p2p layer: the set of connected peers.
+
+    func (sw *Switch) Peers() IPeerSet
+
+The `Peers()` method returns the current set of connected peers.
+The returned `IPeerSet` is an immutable concurrency-safe copy of this set.
+Observe that the `Peer` handlers returned by this method were previously
+[added to the reactor][reactor-addpeer] via the `InitPeer(Peer)` method,
+but not yet removed via the `RemovePeer(Peer)` method.
+Thus, a priori, reactors should already have this information.
+
+    func (sw *Switch) NumPeers() (outbound, inbound, dialing int)
+
+The `NumPeers()` method returns the current number of connected peers,
+distinguished between `outbound` and `inbound` peers.
+An `outbound` peer is a peer the node has dialed to, while an `inbound` peer is
+a peer the node has accepted a connection from.
+The third field `dialing` reports the number of peers to which the node is
+currently attempting to connect, so not (yet) connected peers.
+
+### Broadcast
+
+The switch provides, mostly for historical or retro-compatibility reasons,
+a method for sending a message to all connected peers:
+
+    func (sw *Switch) Broadcast(e Envelope) chan bool
+
+The `Broadcast()` method is not blocking and returns a channel of booleans.
+For every connected `Peer`, it starts a background thread for sending the
+message to that peer, using the `Peer.Send()` method
+(which is blocking, as detailed in [Send Methods](#send-methods)).
+The result of each unicast send operation (success or failure) is added to the
+returned channel, which is closed when all operations are completed.
+
+> _Note_: while this method is helpful, being adopted by several standard
+> reactors, its current _implementation_ is not efficient, as the marshalling
+> of the provided message is performed as part of the `Peer.Send()` helper
+> method, that is, once per connected peer.
+
+### Vetting Peers
+
+The p2p layer relies on the registered reactors to gauge the _quality_ of peers.
+The following method can be invoked by a reactor to inform the p2p layer that a
+peer has presented a "good" behaviour.
+This information is registered in the node's address book and influences the
+operation of the Peer Exchange (PEX) protocol, as node discovery adopts a bias
+towards "good" peers:
+
+    func (sw *Switch) MarkPeerAsGood(peer Peer)
+
+At the moment, it is up to the consensus reactor to vet a peer.
+A peer is marked as good when the consensus protocol receives
+`votesToContributeToBecomeGoodPeer = 10000` useful votes or
+`blocksToContributeToBecomeGoodPeer = 10000` useful block parts from that peer.
```

Comment:
Does the counter decay with time, or is it a matter of time for any non byzantine node to be marked as good?

Reply:
I don't mean the values of `votesToContributeToBecomeGoodPeer` and `blocksToContributeToBecomeGoodPeer` but the counters used to track if the peers have reached these values. But the fact that 
>  The number of peers with this state is limited.
would have a somewhat similar effect.

Reply:
Link to the knowledge-base repository added, I think this discussion should continue there.